lib/ASTGen/Sources/ASTGen/Diagnostics.swift

```diff
@@ -13,65 +13,90 @@
+    )
   }
 
-  var severity: DiagnosticSeverity { .error }
+  /// Emits the given diagnostics via the C++ diagnostic engine.
+  func diagnoseAll(_ diagnostics: [Diagnostic]) {
+    diagnostics.forEach(diagnose)
+  }
 }
 
-/// An error emitted when a token is of an unexpected kind.
-struct UnexpectedTokenKindError: ASTGenError {
-  let token: TokenSyntax
-  private let parent: Syntax
+struct ASTGenDiagnostic: DiagnosticMessage {
+  var node: Syntax
+  var message: String
+  var severity: DiagnosticSeverity
+  var messageID: String
 
-  init(token: TokenSyntax) {
-    guard let parent = token.parent else {
-      preconditionFailure("Expected a child (not a root) token")
-    }
+  var diagnosticID: MessageID {
+    MessageID(domain: "ASTGen", id: messageID)
+  }
 
-    self.token = token
-    self.parent = parent
+  init(node: some SyntaxProtocol, message: String, severity: DiagnosticSeverity = .error, messageID: String) {
+    self.node = Syntax(node)
+    self.message = message
+    self.severity = severity
+    self.messageID = messageID
   }
 
-  var message: String {
-    return """
-      unexpected token kind for token:
-        \(self.token.debugDescription)
-      in parent:
-        \(self.parent.debugDescription(indentString: "  "))
-      """
+  fileprivate init(node: some SyntaxProtocol, message: String, severity: DiagnosticSeverity = .error, function: String = #function) {
+    // Extract messageID from the function name.
+    let messageID = String(function.prefix(while: { $0 != "(" }))
+    self.init(node: node, message: message, severity: severity, messageID: messageID)
   }
 }
 
-/// An error emitted when an optional child token is unexpectedly nil.
-struct MissingChildTokenError: ASTGenError {
-  let parent: Syntax
-  let kindOfTokenMissing: TokenKind
+extension ASTGenDiagnostic {
+  /// An error emitted when a token is of an unexpected kind.
+  static func unexpectedTokenKind(token: TokenSyntax) -> Self {
+    guard let parent = token.parent else {
+      preconditionFailure("Expected a child (not a root) token")
+    }
 
-  init(parent: some SyntaxProtocol, kindOfTokenMissing: TokenKind) {
-    self.parent = Syntax(parent)
-    self.kindOfTokenMissing = kindOfTokenMissing
+    return Self(
+      node: token,
+      message: """
+      unexpected token kind for token:
+        \(token.debugDescription)
+      in parent:
+        \(parent.debugDescription(indentString: "  "))
+      """
+    )
   }
 
-  var message: String {
-    """
-    missing child token of kind '\(self.kindOfTokenMissing)' in:
-      \(parent.debugDescription(indentString: "  "))
-    """
+  /// An error emitted when an optional child token is unexpectedly nil.
+  static func missingChildToken(parent: some SyntaxProtocol, kindOfTokenMissing: TokenKind) -> Self {
+    Self(
+      node: parent,
+      message: """
+      missing child token of kind '\(kindOfTokenMissing)' in:
+        \(parent.debugDescription(indentString: "  "))
```

Comment:
Are these messages intended to be multi-line. I don’t think we have multi-line diagnostics anywhere else at the moment.

Reply:
I think we need to revisit here before enabling ASTGen by default.